File: python/quick_sort.py
# Quick Sort 
# O(nlogn)
# Take integer Array and sort 
# Mahmud Ahsan

def quick_sort(array):
    if len(array) < 2:
        return array
    
    pivot        = array[0] # assume first element as pivot
    less         = [i for i in array if i < pivot]
    greater      = [i for i in array if i > pivot]
    equal        = [i for i in array if i == pivot]
    
    return quick_sort(less) + equal + quick_sort(greater)

#Testing
    
print(quick_sort([1, 2, 4, 4, 5, 1, 3]))
print(quick_sort([5.0, 4.0, 3.0, 2.0, 1.0]))
print(quick_sort(["love", "like", "play", "cool", "my"]))

File: swift/quick_sort.swift
// Quick Sort 
// O(nlogn)
// Take integer Array and sort 
// Mahmud Ahsan

func quickSort <T: Comparable> (_ array: [T]) -> [T] {
    if array.count < 2 {
        return array
    }
    
    let pivot    = array[0] // assume first element as pivot
    let less     = array.filter { $0 < pivot  }
    let greater  = array.filter { $0 > pivot  }
    let equal    = array.filter { $0 == pivot }
    
    return quickSort(less) + equal + quickSort(greater)
}

// Testing
assert([1, 1, 2, 3, 4, 4, 5] == quickSort([1, 2, 4, 4, 5, 1, 3]))
assert([1.0, 2.0, 3.0, 4.0, 5.0] == quickSort([5.0, 4.0, 3.0, 2.0, 1.0]))
assert(["cool", "like", "love", "my", "play"] == quickSort(["love", "like", "play", "cool", "my"]))